aten/src/ATen/TensorImpl.h

```diff
@@ -121,26 +116,81 @@ struct AT_API TensorImpl : public Retainable {
 
   // represents that numel() == 0.
   inline bool is_empty() const {
-    for (int64_t i = 0; i < dim(); ++i) {
-      if (sizes()[i] == 0) {
-        return true;
-      }
-    }
-    return false;
+    return numel() == 0;
+  }
+
+  virtual void resize_dim(int64_t ndim) {
+    // NB: This is *truly* a resize; calling code (e.g., squeeze)
+    // assumes that old values are preserved
+    sizes_.resize(ndim);
```

Comment:
why not have a protected virtual _sizes_mut() / _strides_mut() so you can implement all these functions once non virtually?

Reply:
It's more difficult to keep track of sizes_ if it's been referenced somewhere, then implement these things (or disable them explicitly) for subclasses. Also, we're going to devirtualize all the functions later on anyway right? And even further along, this will be merged into Tensor.